### PR TITLE
drawEffect delta time refactor

### PR DESCRIPTION
**Pull request by EvidentlyCube**
Feel free to review the code of individual effects, I don't think it'll be changing anymore, maybe some specific effects that are broken if any are broken. 

The rest is still WIP.

### CODE REVIEW

**Review by mrimer — COMMENTED** [review]
Really nice cleanup!

I posted some areas that might need attention, plus some stylistic suggestions.

A couple other points:
* Not trying to give you more work, but it might make sense to refactor the declaration of the dwDuration member var from multiple subclasses of CEffect into CEffect itself.
* With the FrontEndLib changes, drodrpg effects need to be updated as well to avoid breakage, correct?

**Comment by EvidentlyCube** [comment]
Okay, that was unexpected amount of work and I might have went overboard. It's past midnight so I can't really write much, but let me quickly summarize things:

- I was pretty much done with the changes when I realized that there is a chicken and egg problem with lighting and image overlays on O layer. Essentially if you have an image overlay that moves on that layer, it'd not be rendered with proper lighting on any new tile it entered this frame, because floor there was already drawn with lighting, and when T layer is drawn on top of it it would not apply lighting because it was already applied on O layer. 
- Then I read your suggestion about making dwDuration common and I figured that solving the above problem would require knowing the dirty rects of the image overlay before it's drawn - but it's not possible with how Draw handles everything.
- So I split each effect processing into two stages - update and draw.
- **First let me say - it's not finished and it's crashing**. 5 minutes ago I finished all changes and made it build without compilation errors or warnings, so I'll likely need another evening to remove crashes and maybe another to ensure all effects behave as they behaved before.
- We might ultimately not need the Update/Draw split for anything other than the problem I noted at the beginning, I do think it's cleaner and ultimately it lead to removing a lot of code duplication. Probably more would be possible but I tried to modify effects as little as possible.
- It was fun and horrifying, I hope you won't say "meh, let's ditch it" but at the same time I am open to any improvements.
- You may want to wait until the end of the weekend to review the code, or  if you have a lot of free time you don't know how to spend better feel free to start biting into it :),

**Comment by EvidentlyCube** [comment]
I've pushed a first batch of changes, fixing stuff that I have managed to quickly find to be broken.

Next I'll check that every effect in the game runs correctly and fix any issues that crop up.

**Comment by EvidentlyCube** [comment]
And checked all the effects:

```
                               REGULAR
AnimatedTileEffect                 <only ever extended by other effects>    
AumtlichGazeEffect                  ✔
BloodEffect                         ✔
BumpObstacleEffect                  ✔
CheckpointEffect                    ✔
DebrisEffect                        ✔
EvilEyeGazeEffect (temporary)       ✔ 
EvilEyeGazeEffect (static)          ✔
ExpandingTextEffect                 ✔
ExplosionEffect                     ✔
FadeTileEffect                      ✔ (not used in DROD but tested it anyway)
FiretrapEffect                      ✔
FlashMessageEffect                  ✔
FlashShadeEffect                    ✔ (not used in DROD or RPG but tested)
FloatEffect                         ✔
FloatTextEffect                     ✔ (not used in DROD by tested it anyway)
FrameRateEffect                     ✔
IceMeltEffect                       ✔
ImageOverlayEffect                  ✔ (tested a couple of more complex cases + TSS epilogue + TSS whirlwinds + TSS tuning fork rooms)
MovingTileEffect                    ❌ (Not used anywhere and it's broken, but looking at the code it seems like it was broken long time ago)
NoticeEffect                        ✔
ParticleExplosionEffect             ✔
PendingBuildEffect                  ✔
PendingPlotEffect                   ✔
PuffExplosionEffect                 ✔
RaindropEffect                      ✔
RoomDrawStatsEffect                 ✔
RotateTileEffect                    ✔ (not used in DROD or RPG but tested)
ShadeEffect                         ✔
SnowflakeEffect                     ✔
SparkEffect                         ✔
SpikeEffect                         ✔
SplashEffect                        ✔
StrikeOrbEffect                     ✔
StunEffect (fading)                 ✔
StunEffect (turn-based)             ✔
SubtitleEffect                      ✔ - Small behavior change: now when new line of text is added to a subtitle effect, if it has already faded out by more than 50% it will replace the whole textbox
SwordsmanSwirlEffect                ✔
SwordSwingEffect                    ✔
TemporalMoveEffect                  ✔
TrapdoorFallEffect                  ✔
VarMonitorEffect                    ✔
VerminEffect                        ✔
WadeEffect                          ✔
```

**Comment by EvidentlyCube** [comment]
I am converting it from draft to proper PR now

File: DROD/EditSelectScreen.cpp [hunk 1]
```diff
@@ -2556,7 +2556,7 @@ void CEditSelectScreen::PopulateHoldListBox()
 #ifdef ENABLE_CHEATS
 				false
 #else
-				!g_pTheDB->Holds.PlayerCanEditHold(pHold->dwHoldID)
+				false//!g_pTheDB->Holds.PlayerCanEditHold(pHold->dwHoldID)
```
Comment (mrimer) [inline DROD/EditSelectScreen.cpp:2559]: This change doesn't look right.  Maybe a dev tweak that needs to be backed out?

File: DROD/PendingPlotEffect.cpp [hunk 2]
```diff
@@ -142,7 +175,7 @@ void CPendingPlotEffect::PlotSwordsman(
 //
 //Params:
 	const UINT wStartX, const UINT wStartY, const UINT wEndX, const UINT wEndY,
-	SDL_Surface* pDestSurface)
+	SDL_Surface& pDestSurface)
```
Comment (mrimer) [inline DROD/PendingPlotEffect.cpp:178]: Probably should rename pDestSurface --> destSurface wherever the type changed from a pointer to a reference.

File: DROD/RoomWidget.cpp [hunk 3]
```diff
@@ -2361,18 +2364,22 @@ void CRoomWidget::RenderRoomLayers(SDL_Surface* pSurface, const bool bDrawPlayer
 	ASSERT(this->pRoom);
 
 	RenderFogInPit(pSurface);
-	this->pOLayerEffects->DrawEffectsOfType(EIMAGEOVERLAY, false, true, pSurface);
+	// @FIXME-FREEZING
```
Comment (mrimer) [inline DROD/RoomWidget.cpp:2367]: To address (or clarify what's the issue in the comment, and why we can defer changes till later).

**Review by mrimer — CHANGES_REQUESTED** [review]
This is a really nice refactor and cleanup!

I get the decoupling of the Update/Draw portions, and it's so much easier to read through any particular effect now.

I should note you probably understand the orb bolt effect code better than anyone right now.

Thank you for making these changes and for the comprehensive testing of all effects.

I've noted a few things:
* potential fix required
* where some additional cleanup would be useful to remove vestiges of the Old Way.
* need for additional comments to address the TOFIX areas you've noted.

**Comment by mrimer** [comment]
@EvidentlyCube , careful about updating effects in FrontEndLib.
Many of these are used in the DROD RPG codebase as well, e.g., CMovingTileEffect.  I didn't realize this one was broken for a long time; it should be fixed up if there's an issue with it.

Also, we need to update the effect classes in the RPG codebase to work with these changes.

File: DROD/EditSelectScreen.cpp (hunk 1 above)
Comment (EvidentlyCube) [reply DROD/EditSelectScreen.cpp:2559]: Yup, accidentally committed debug code.

File: DROD/PendingPlotEffect.cpp (hunk 2 above)
Comment (EvidentlyCube) [reply DROD/PendingPlotEffect.cpp:178]: You asked for it - 73 files updated :).

File: DROD/RoomWidget.cpp (hunk 3 above)
Comment (EvidentlyCube) [reply DROD/RoomWidget.cpp]: Those are not needed anymore, removed

**Comment by EvidentlyCube** [comment]
Next steps:

 - There is still a small visual bug with image overlay on Layer 0 with rooms with darkness, there is one frame blink without shading if you playtest in such a room. Feel free to cehck the attached hold, playtest in 1N: [aaa.zip](https://github.com/CaravelGames/drod/files/5442036/aaa.zip)
 - Get RPG to build (I'll hope it'll just magically work thanks to the clever script) and fix any compilation errors that might be there.
 - Okay, you're right with `MovingTileEffect`. I could swear there was another effect that I tested and it worked that was used by RPG but I must've mixed something up. I'll fix it as part of the above, I might've been wrong with my diagnosis that it was broken for a long time!

File: DROD/PendingPlotEffect.cpp (hunk 2 above)
Comment (mrimer) [reply DROD/PendingPlotEffect.cpp]: Thank you. I really appreciate the care in making it easy to read!

**Comment by mrimer** [comment]
Thanks for the updates.  They all look good!

I agree on what's left to do.  Looking forward to it!

**Comment by EvidentlyCube** [comment]
And pushed final changes which introduce a lot of, erm, changes:

1. Ported effect changes to RPG
2. Built RPG and checked effects - I wasn't as thorough as with regular DROD so things might not work fully as intended.
3. A lot of changes in how lighting is applied in `CRoomWidget`. Check the commits for details!

**Comment by EvidentlyCube** [comment]
image

File: DROD/RoomWidget.cpp [hunk 4]
```diff
@@ -4420,6 +4400,11 @@ void CRoomWidget::Paint(
 	if (bMoveAnimationInProgress && ShowShadowCastingAnimation())
 		AddPlayerLight(true);
 
+	// O-Layers need to be updated and dirties before O-Layer is drawn to ensure lighting is correctly applied for
```
Comment (mrimer) [inline DROD/RoomWidget.cpp:4403]: dirties -> dirtied

**Review by mrimer — APPROVED** [review]
I cannot overstate the goodness of this PR.  Thank you!

**Comment by EvidentlyCube** [comment]
Updated!

**Comment by mrimer** [comment]
@EvidentlyCube , I really like the design changes you've made in this PR.  I admit I can't follow the intricacies of the redesign of layering and lighting in RoomWidget precisely merely by perusing the diff, but I like how the new screenshots appear, and everything in the latest code changes seems reasonable and there's nothing obviously amiss that I can see.  Your testing was pretty comprehensive, and of course we can roll forward with any regression fixes that are uncovered in the future.

Just to ensure we're on the same page, do you consider this PR ready to merge?

**Comment by EvidentlyCube** [comment]
@mrimer Yes, the PR is redy to merge. I wouldn't say I was super comprehensive with testing but I tried to be as much as I could while remaining sane.

**Comment by TomBrouws** [comment]
It seems DROD is crashing for me since these changes have been made, the line mentioned in the crash report is ImageOverlayEffect.cpp:334. I can consistently trigger the crash by selecting 'DROD Touch' in the Hold Selection screen on my main player profile, which has DROD Touch: World 3: Once South, Once East open as the current room.

I can try to investigate but @EvidentlyCube is probably much better equipped to do that.

Full stack trace:
```
Termination Signal:    Segmentation fault: 11
Termination Reason:    Namespace SIGNAL, Code 0xb
Terminating Process:   exc handler [62758]

Thread 0 Crashed:: Dispatch queue: com.apple.main-thread
0   drod                          	0x000000010c9d97ed CImageOverlayEffect::UpdateCommand(ImageOverlayCommand const&, CommandExecution&, unsigned int) + 27 (ImageOverlayEffect.cpp:334)
1   drod                          	0x000000010c9d9482 CImageOverlayEffect::AdvanceState(unsigned int) + 102 (ImageOverlayEffect.cpp:248)
2   drod                          	0x000000010c9d8e3b CImageOverlayEffect::Update(unsigned int, unsigned int) + 55 (ImageOverlayEffect.cpp:151)
3   drod                          	0x000000010ca57442 CEffect::Update(unsigned int) + 92
4   drod                          	0x000000010ca57fa5 CEffectList::UpdateEffect(CEffect*, bool) + 113
5   drod                          	0x000000010ca57d92 CEffectList::UpdateEffects(bool, unsigned int) + 160
6   drod                          	0x000000010c9b080d CRoomWidget::Paint(bool) + 1487 (RoomWidget.cpp:4387)
7   drod                          	0x000000010ca48bf1 CWidget::PaintChildren(bool) + 291
8   drod                          	0x000000010c8f9de8 CGameScreen::Paint(bool) + 114 (GameScreen.cpp:5430)
9   drod                          	0x000000010ca533f9 CScreenManager::ActivateScreen(unsigned int) + 371```

**Comment by EvidentlyCube** [comment]
Let's see if reply by mail works.

It's most likely caused by accessing vector outside its bounds - I fixed it
here
https://github.com/CaravelGames/drod/pull/274/files#diff-94da852cd53643a3c3e77ae3e51ab219e9977c0d53a6fb64d5d0926e56517411

Can you try on latest master?

**Pull request merged**